### PR TITLE
[xaml] Handle deeply nested translation attributes, allow translatable values to be expressed as a descendent element instead of attribute

### PR DESCRIPTION
**Pull request by adamint**
The purpose of this PR is to allow:
- attributes/elements deeply nested (>2 parents) to be expressed as a translatable node with a descriptive id, and to
- allow translatable nodes to be expressed as a descendent element instead of an attribute value. see below example:

```xml
<A Name="MyName">
            <A.Description>MyName Description</A.Description> <!-- this is translated -->
            <B>
              <C>
                <D>
                  <NameValuePair Name="SearchTerms" TranslatableProp1="this-is-translated" xliff:LocalizedProperties="TranslatableProp1"> <!-- TranslatableProp1 is a translated attribute, while NameValuePair.Value is a translated descendent element -->
                    <NameValuePair.Value>My Search Terms</NameValuePair.Value> 
                </D>
                ....
 ```

### CODE REVIEW

File: src/XliffTasks/Model/XamlRuleDocument.cs [hunk 1]
```diff
@@ -57,25 +118,75 @@ protected override IEnumerable<TranslatableNode> GetTranslatableNodes()
...
+            if (ancestorWithNameAttributeCandidate is null)
+            {
+                if (element.Parent is not null)
+                {
+                    idBuilder.Append(element.Parent.Attribute("Name") is null ? XmlName(element.Parent) : $"{XmlName(element.Parent)}|{AttributedName(element.Parent)}");
+                    idBuilder.Append("|Metadata|");
+                }
+
+                idBuilder.Append(XmlName(element));
+                if (element.Attribute("Name") is null)
```
Comment (adamint) [inline src/XliffTasks/Model/XamlRuleDocument.cs:158]: this is nearly duplicated code, I will update

File: src/XliffTasks.Tests/XamlRuleDocumentTests.cs [hunk 2]
```diff
@@ -10,6 +10,524 @@ namespace XliffTasks.Tests
...
+</Rule>
+""";
+        
+        XamlRuleDocument document = new();
+        StringWriter writer = new();
+        document.Load(new StringReader(source));
+
+        XlfDocument xliffDocument = new();
+        xliffDocument.LoadNew("fr");
+        xliffDocument.Update(document, "test.xaml");
```
Comment (drewnoakes) [inline src/XliffTasks.Tests/XamlRuleDocumentTests.cs:526]: What is this test covering? The input is so huge it's hard to see what task it's performing, and there doesn't seem to be any expected output.

File: src/XliffTasks.Tests/XamlRuleDocumentTests.cs [hunk 3]
```diff
@@ -160,33 +678,205 @@ public void BasicLoadAndTranslate()
...
+      <ValueEditor EditorType=""MultiStringSelector"">
+        <ValueEditor.Metadata>
+          <NameValuePair Name=""TypeDescriptorText"" Value=""Custom symbols"" xliff:LocalizedProperties=""Value"" />
+          <NameValuePair Name=""SearchTerms"">
+            <NameValuePair.Value>My search terms</NameValuePair.Value>
+          </NameValuePair>
+
+          <TestElement Name=""TranslatableElement"" LocalizablePropC=""LocalizableC attribute"" xliff:LocalizedProperties=""LocalizablePropA;LocalizablePropB;LocalizablePropC"">
+            <TestElement.LocalizablePropA>LocalizableA descendent prop</TestElement.LocalizablePropA>
+            <TestElement.LocalizablePropB>LocalizableB descendent prop</TestElement.LocalizablePropB>
```
Comment (drewnoakes) [inline src/XliffTasks.Tests/XamlRuleDocumentTests.cs:738]: Technically, property element syntax can be declared on a base type. For example, if class `C` derives from `Base`, and `Base` defines property `Property`:

```xml
<C>
  <Base.Property>Hello</Base.Property>
<C>
```

is equivalent to:

```xml
<C>
  <C.Property>Hello</C.Property>
<C>
```

We don't understand the type hierarchy of values defined here, so we can't easily obtain total parity. Options are to 1) leave it as is, 2) allow `*.Property` without requiring stuff before the `.` to match the parent.

File: src/XliffTasks.Tests/XamlRuleDocumentTests.cs [hunk 4]
```diff
@@ -160,33 +678,205 @@ public void BasicLoadAndTranslate()
...
+                    <NameValuePair.Value>My Search Terms</NameValuePair.Value> 
+                </D>
+                ....
+    
+        */
+        [Fact]
+        public void LoadAndTranslateWithDeeplyNestedTranslationAndElementAttributeSyntax()
+        {
+            string source =
+@"<Rule Name=""MyRule""
```
Comment (drewnoakes) [inline src/XliffTasks.Tests/XamlRuleDocumentTests.cs:712]: `"""` style comments would make this test data slightly easier to read. Ditto elsewhere.

File: src/XliffTasks/Model/XamlRuleDocument.cs [hunk 5]
```diff
@@ -22,29 +24,88 @@ protected override IEnumerable<TranslatableNode> GetTranslatableNodes()
         {
             foreach (XElement? element in Document.Descendants())
             {
+                // first, let's check if the element has a descendent by the name of {elementLocalName.Description/DisplayName}
+                var descendentDisplayName = element.Descendants(XName.Get($"{element.Name.LocalName}.DisplayName", element.Name.NamespaceName)).FirstOrDefault();
+                var descendentDescription = element.Descendants(XName.Get($"{element.Name.LocalName}.Description", element.Name.NamespaceName)).FirstOrDefault();
```
Comment (drewnoakes) [inline src/XliffTasks/Model/XamlRuleDocument.cs:29]: This should only be searching children, not all descendants.

```suggestion
                var descendentDisplayName = element.Elements(XName.Get($"{element.Name.LocalName}.DisplayName", element.Name.NamespaceName)).FirstOrDefault();
                var descendentDescription = element.Elements(XName.Get($"{element.Name.LocalName}.Description", element.Name.NamespaceName)).FirstOrDefault();
```

File: src/XliffTasks/Model/XamlRuleDocument.cs [hunk 6]
```diff
@@ -22,29 +24,88 @@ protected override IEnumerable<TranslatableNode> GetTranslatableNodes()
...
+                }
+                
+                var localizableProperties = element.Attribute(XName.Get(LocalizedPropertiesAttributeName, XliffTasksNs))?.Value?.Split(';');
+
+                if (localizableProperties is not null)
+                {
+                    // we could have any number of descendent localizable properties
+                    foreach (var localizableProperty in localizableProperties)
+                    {
+                        if (element.Descendants(XName.Get($"{element.Name.LocalName}.{localizableProperty}", element.Name.NamespaceName)).FirstOrDefault() is { } descendentValue)
```
Comment (drewnoakes) [inline src/XliffTasks/Model/XamlRuleDocument.cs:58]: I wonder if this should only search children, not all descendants.

File: src/XliffTasks/Model/XamlRuleDocument.cs [hunk 7]
```diff
@@ -22,29 +24,88 @@ protected override IEnumerable<TranslatableNode> GetTranslatableNodes()
...
+                        if (XmlName(attribute) == "Value")
+                        {
+                            yield return new TranslatableXmlAttribute(
+                                id: GenerateIdForPropertyMetadata(element),
+                                source: attribute.Value,
+                                note: GetComment(element, XmlName(attribute)),
+                                attribute: attribute);
+                        }
+                        // else if we have a descendent in the form of {elementLocalName}.Value, we should translate that descendent
+                        else if (element.Descendants(XName.Get($"{element.Name.LocalName}.Value", element.Name.NamespaceName)).FirstOrDefault() is { } descendentValue)
```
Comment (drewnoakes) [inline src/XliffTasks/Model/XamlRuleDocument.cs:91]: Ditto, search children, not descendants.

**Pull request merged**